chain/testing.go

```diff
@@ -31,7 +31,9 @@ type Builder struct {
 	stateBuilder StateBuilder
 	seq          uint64 // For unique tickets
 
-	blocks map[cid.Cid]*types.Block
+	blocks   map[cid.Cid]*types.Block
+	messages map[cid.Cid][]*types.SignedMessage
+	receipts map[cid.Cid][]*types.MessageReceipt
```

Comment:
Any messages/receipts created in the `build` function below should be stored here at the same time as the block is stored. That will need a little further refactoring when we split the messages out.

I would prefer that the `StoreXxx` methods below did not exist, in favour of this mechanism. For things that want a (fake) message store but will not be using messages attached to blocks built by this builder, just use the MessageStore, or a distinct FakeMessageStore with these maps.

Reply:
I also would prefer it this way.  If anyone has ideas I could use some help.

Reply:
Current idea: private fields on block builder share messages and receipts cache.  The chain builder passes them in during builder construction.  `AddMessage` / `AddReceipt` callbacks defined in chain add messages / receipts directly.  Out of package callers can't mess up setup because builder has control of which caches to share.  I think this should work pretty well, speculatively including it in this PR.